### ISSUE TITLE
Extend `test` command to allow filtering package / file / test

### ISSUE THREAD

**Issue by anderseknert** [open-policy-agent/opa#6696]
Having recently looked into a deeper integration of the OPA test runner (via command line invocation) and the [OPA VS Code extension](https://marketplace.visualstudio.com/items?itemName=tsandall.opa), there are some hurdles that I think would be best solved in OPA itself, as that would benefit not just VS Code but any editor including a similar integration. The problems are described in [this issue](https://github.com/open-policy-agent/vscode-opa/issues/136), but to summarize here:

Editors and IDEs commonly run tests filtered by either:

- file (run all tests in file) 
- package (run all tests in package)
- name _inside of a file_ (click test to run)

The `opa test` command doesn't really accomodate any of the above needs, as it'll always run _all_ tests in the files that are loaded. From an editors POV, that'll be all files in the workspace, or all Rego files really. It's possible to workaround this in editors by actually running all the tests, but only presenting the filtered result to the user, making them think that only the tests they asked for got run. It's also possible to use the `--run` argument to pass a list of regular expressions that should match the name of the tests, but that doesn't take the name of the file or package into account... and a similarly named test could theoretically exist in any number of files without them being logically related.

Suggestion here is to either extend what `--run` accepts, so that it would be possible to provide something like `package/*`, `file.rego/*` or `file.reg/test_1`, or to introduce another flag for this purpose. Naturally, this would need to work as expected with things like coverage, etc.

**Comment by charlieegan3** [comment]
I'd like to work on getting this done for some client improvements.

Thinking aloud, `--ignore` and `--bundle` are not the right way to do this since they alter what is loaded, not what is run. We want them to work as usual.

> It's also possible to use the --run argument to pass a list of regular expressions

Kind of, it's still a single regular expression and extensions to this format to support files and packages sound pretty unfriendly / error prone to me.

```sh
opa test . --run "test_foo|test_bar"
# this might be nicer?
# opa test . --run test_bar --run test_foo
```

What about two new flags:

* `--run-package` works the same way, but matches based on package.
* `--run-file` this might be better as a glob match of some kind supporting things like `dir/*` and matching `dir/file.rego` with just `file.rego` etc. Less sure about the implementation of that one.

Would that do the job?

**Comment by anderseknert** [comment]
Looking at this issue now, and I believe this predates Regal having a language server implementation. Approaching this today, I don't think I'd want to have the VS code test runner shell out to the `opa test` command, but trigger a command on the server that runs requested tests and returns the results. The issues mentioned here are still valid though, just that we'd want that kind of filters passed to a Go API rather than via command line flags. Obviously, we can have those added in OPA (and `regal test`)  too using the same API.

However, neither the test runner [nor the benchmark feature](https://github.com/open-policy-agent/opa/issues/8225) is currently exposed via public APIs! So we can't really do much without first adding programmatic entrypoints for testing in OPA. That's obviously a bigger undertaking than extending the `opa test` command! But as the linked benchmarking issue demonstrates, something we'll want for Regal anyway. What do you think?

**Comment by charlieegan3** [comment]
Yeah that makes the most sense. I think we can do some tidying up in the extensions first to avoid needing OPA at all, and then build out the testing functionality.
